feat(exomint): introduce module

MaxMustermann2 opened the pull request:
Includes #102 and should ideally be reviewed after that one is merged.

The `x/exomint` module is a module unique to the Exocore chain. It is used to produce the annual staking-related inflation. Compared to the Cosmos SDK's `mint` module, the `exomint` module mints the rewards only at the end of an epoch, while the Cosmos SDK produces rewards at each block. The number of coins minted differs in the two as well - while Cosmos decides the number based on a target inflation rate and the currently bonded ratio, `x/exomint` simply mints a fixed amount of native tokens at each epoch (this design is **TBD**).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced the `exomint` module with new epoch-related functionalities.
  - Added CLI commands for querying and managing epoch information in the `epochs` module.
  - Implemented new genesis state functions for the `epochs` module.

- **Bug Fixes**
  - Updated import paths to reflect new module locations.

- **Tests**
  - Added comprehensive test scenarios for epoch management and block production.

- **Documentation**
  - Updated README for the `epochs` module.
  - Added .semgrepignore rules for improved code scanning.

- **Chores**
  - Updated GitHub workflows with new environment settings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
feat(exomint): introduce module

The exomint module is a variation of the Cosmos SDK's mint module. The
mint module depends mandatorily on the staking token supply and
optionally on the bonded ratio of the token to calculate the amount of
tokens to be minted in the block. Since none of these two metrics are
super meaningful to us in our restaking environment, we are using a
custom block reward minting module.

The module requires two parameters as input: the `MintDenom` which
defaults to the default bond denomination of the chain and the
`BlockReward` which defaults to 1. In other words, by default, this
module mints 1 stake denomination worth of coins at the beginning of the
block to the fee collector address.

The module name was chosen as `exomint` to ensure that it does not clash
with the SDK's module name (for clarity) while keeping go package naming
conventions (no underscores) in mind.

MaxMustermann2 committed:
chore: update go version to 1.21.11

MaxMustermann2 committed:
Merge develop into exomint

MaxMustermann2 committed:
refactor(exomint): epoch minting

MaxMustermann2 committed:
fix: do not use hardcoded param

MaxMustermann2 committed:
chore: golang lint

MaxMustermann2 committed:
chore: lint proto

MaxMustermann2 committed:
test(exomint): fix tests

MaxMustermann2 committed:
feat(exomint): allow update params

The function is currently guarded by the address of operator2. It should
be switched over to the governance address when it is ready.

MaxMustermann2 committed:
fix(exomint): stateful validation add

MaxMustermann2 committed:
fix(ci): ignore operator address line

MaxMustermann2 committed:
fix(exomint): add tx to cli

MaxMustermann2 committed:
feat(epochs): backport the epochs module

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
fix: run go mod tidy

MaxMustermann2 committed:
fix(local): minute-wise epochs

MaxMustermann2 committed:
fix(dogfood): set default unbonding duration 7 day

MaxMustermann2 committed:
fix(local): unbond at 2 epochs

MaxMustermann2 committed:
fix(ci): fix failing jobs

MaxMustermann2 committed:
refactor(epochs): respond to ai review

MaxMustermann2 committed:
fix(docs): add `the` before module

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
epochs: return block time in query

The returned block time can be used to determine whether any of the
epochs are about to change.

MaxMustermann2 committed:
doc(epochs): clarify no problems with timezones

MaxMustermann2 committed:
fix(ci): provide github token to buf setup

avoid rate limits

MaxMustermann2 committed:
feat(ci): allow super-linter running with act

Example usage:
act --workflows .github/workflows/super-linter.yml pull_request

MaxMustermann2 committed:
Merge with `develop`

MaxMustermann2 committed:
remove unused file

MaxMustermann2 committed:
fix: typo in script

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
fix(app): revert authority to gov address

The hardcoded authority is not a good design.

MaxMustermann2 committed:
doc(exomint): update comment about params

Not all params need to be supplied. Only the ones being edited. Those
that are not supplied are loaded from the previous parameter set.

MaxMustermann2 committed:
fix(exomint): return correct type from tx

leonz789 commented on x/exomint/keeper/params.go:
```diff
@@ -0,0 +1,60 @@
+// GetMintDenom gets the mint denomination.
+func (k Keeper) GetMintDenom(ctx sdk.Context) string {
+	var mintDenom string
+	k.paramstore.Get(ctx, types.KeyMintDenom, &mintDenom)
```
x/params has been depreacted ,and for the gov module to handle params update through proposal can just by setting the authority with govModule(it has registered the msgServiceRouter with all modules in its keeper). Maybe consider just use module store to house the params ?

MaxMustermann2 replied:
Done, thank you

leonz789 commented on x/exomint/types/params.go:
```diff
@@ -0,0 +1,101 @@
+	"gopkg.in/yaml.v2"
+)
+
+var _ paramtypes.ParamSet = (*Params)(nil)
```
we can just use module store instead of x/parmas since it's depreacted.

MaxMustermann2 committed:
fix(exomint): do not use deprecated x/params

MaxMustermann2 commented:
I tested this out by setting the minting epoch to `minute` in `local_node.sh` and of course defining the minute epoch.

```
$ exocored query bank balances exo17xpfvakm2amg962yls6f84z3kell8c5lgqz40d
balances: []
pagination:
  next_key: null
  total: "0"
$ exocored query bank balances exo17xpfvakm2amg962yls6f84z3kell8c5lgqz40d
balances:
- amount: "20000000000000000000"
  denom: aexo
pagination:
  next_key: null
  total: "0"
```

MaxMustermann2 committed:
test(exomint): add some tests

Currently remaining tests are ABCI test, msg test and CLI msg test

MaxMustermann2 committed:
Merge port-epochs

cloud8little commented:
Seems the fixed amount of native tokens can not be changed after the network runs. It may be better to update the fixed amount by some multiple-signature account.

MaxMustermann2 committed:
update app for port-epochs

MaxMustermann2 committed:
fix(exomint): update for port-epochs

MaxMustermann2 committed:
feat(exomint): emit epoch id in event

MaxMustermann2 committed:
test(exomint): add epoch hook test

MaxMustermann2 committed:
fix(exomint): validate params before update

MaxMustermann2 commented:
> Seems the fixed amount of native tokens can not be changed after the network runs. It may be better to update the fixed amount by some multiple-signature account.

You can change it with `MsgUpdateParams` via governance. If needed we can remove the governance gating check. However, that will run afoul of #90

MaxMustermann2 committed:
chore(exomint): gofumpt

MaxMustermann2 committed:
fix(exomint): validate overriden params

MaxMustermann2 committed:
fix(exomint): make tests pass

MaxMustermann2 commented on x/exomint/keeper/params.go:
```diff
@@ -0,0 +1,25 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// GetParams get all parameters as types.Params
+func (k Keeper) GetParams(ctx sdk.Context) types.Params {
+	store := ctx.KVStore(k.storeKey)
+	key := types.KeyPrefixParams()
+	bz := store.Get(key)
+	var params types.Params
+	k.cdc.MustUnmarshal(bz, &params)
+	return params
+}
+
+// SetParams set the params
+func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.KeyPrefixParams()
+	bz := k.cdc.MustMarshal(&params)
+	store.Set(key, bz)
+}
```
That is intentional. A failure in de/serialization is a panic-worthy issue.

MaxMustermann2 commented on x/exomint/keeper/impl_epochs_hooks_test.go:
```diff
@@ -0,0 +1,39 @@
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+)
+
+func (suite *KeeperTestSuite) TestEpochHooks() {
+	epsilon := time.Nanosecond // negligible amount of buffer duration
+	params := suite.App.ExomintKeeper.GetParams(suite.Ctx)
+	feeCollector := suite.App.AccountKeeper.GetModuleAddress(
+		authtypes.FeeCollectorName,
+	)
+	// default is day, we start by committing after 1 minute
+	suite.SetupTest() // reset
+	suite.CommitAfter(time.Minute + epsilon)
+	// check balance
+	suite.Require().True(
+		suite.App.BankKeeper.GetBalance(
+			suite.Ctx,
+			feeCollector,
+			params.MintDenom,
+		).Amount.Equal(sdkmath.NewInt(0)),
+	)
+
+	// now go to one day
+	suite.CommitAfter(time.Hour*24 + epsilon - time.Minute)
+	// check balance
+	suite.Require().True(
+		suite.App.BankKeeper.GetBalance(
+			suite.Ctx,
+			feeCollector,
+			params.MintDenom,
+		).Amount.Equal(params.EpochReward),
+	)
+}
```
Any more comments will be too verbose. I am happy with the current set.

MaxMustermann2 commented on x/exomint/keeper/impl_epochs_hooks.go:
```diff
@@ -0,0 +1,78 @@
+func (wrapper EpochsHooksWrapper) AfterEpochEnd(
+	ctx sdk.Context, identifier string, number int64,
+) {
+	params := wrapper.keeper.GetParams(ctx)
+	if strings.Compare(identifier, params.EpochIdentifier) == 0 {
+		mintedCoin := sdk.NewCoin(
+			params.MintDenom,
+			params.EpochReward,
+		)
+		mintedCoins := sdk.NewCoins(mintedCoin)
+
+		logger := wrapper.keeper.Logger(ctx)
+		err := wrapper.keeper.MintCoins(ctx, mintedCoins)
+		if err != nil {
+			logger.Error(
+				"AfterEpochEnd",
+				"could not mint coins", err,
+			)
+			return
+		}
+
+		err = wrapper.keeper.AddCollectedFees(ctx, mintedCoins)
+		if err != nil {
+			logger.Error(
+				"AfterEpochEnd",
+				"could not transfer coins", err,
+			)
+			return
+		}
```
If the suggestion is to replace `Could not mint coins: $err` with `Failed to mint coins due to $err` it is purely semantics and not necessary because the spirit of the two lines is the same.

MaxMustermann2 committed:
fix(exomint): `post` the UpdateParams

MaxMustermann2 commented on proto/exocore/exomint/v1/tx.proto:
```diff
@@ -0,0 +1,36 @@
+  rpc UpdateParams(MsgUpdateParams) returns (MsgUpdateParamsResponse) {
+    option (google.api.http).get = "/exocore/exomint/v1/tx/MsgUpdateParams";
+  }
+}
```
Fixed in ae52514f

cloud8little committed:
Merge branch 'develop' into exomint

cloud8little commented:
Test passed.

MaxMustermann2 committed:
fix(test): change localnet minting frequency

mikebraver commented on x/exomint/keeper/keeper.go:
```diff
@@ -0,0 +1,76 @@
+// MintCoins to be used in BeginBlocker.
+func (k Keeper) MintCoins(ctx sdk.Context, newCoins sdk.Coins) error {
+	if newCoins.Empty() {
+		// skip as no coins need to be minted
```
Would this happen? As if there's no coins need to be minted, will this function be called?

MaxMustermann2 replied:
It is a bit of an excessive short-circuit mechanism. I have removed it in the next commit.

bwhour commented on local_node.sh:
```diff
@@ -128,6 +128,11 @@ if [[ $overwrite == "y" || $overwrite == "Y" ]]; then
 	jq '.app_state["dogfood"]["last_total_power"]="5000"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
 
+	# x/exomint
+	# the default epoch_identifier is day, which means that rewards are minted once every day
```
the comment said the default is **day**, but the code is **minute**.

MaxMustermann2 replied:
The comment talks about the default identifier (day) which the code changes (to minute)

MaxMustermann2 replied:
Will update the comment with some more clarity.

bwhour commented on x/exomint/types/params.go:
```diff
@@ -0,0 +1,125 @@
+	return sdk.ValidateDenom(v)
+}
+
+func ValidateEpochReward(i interface{}) error {
```
there is one case in this function:
need it  add an upper limit check to prevent excessive reward amounts ?

MaxMustermann2 replied:
Given that this number is set during genesis and can only be modified by governance, such a check can be skipped.

bwhour reviewed (approved):
it looks good.

MaxMustermann2 committed:
fix(local): clarify comment about exomint

The default epoch identifier set in the module is `day`. The script
changes it to `minute` to enable easier localnet testing

MaxMustermann2 committed:
fix(exomint): support 0 value of rewards

In the future, a possibility that governance decides to turn off annual
inflation exists. To that end, the `exomint` module should not reject a
value of 0 for the epoch reward.

MaxMustermann2 commented:
Test object: 149821af0e4e8877ecf16fddb3d16c0f4f660978

With patch for oracle module
```diff
diff --git a/local_node.sh b/local_node.sh
index 48e05aa2..d7e7c90f 100755
--- a/local_node.sh
+++ b/local_node.sh
@@ -133,6 +133,9 @@ if [[ $overwrite == "y" || $overwrite == "Y" ]]; then
 	# which is more suitable for a live network and not a localnet.
 	jq '.app_state["exomint"]["params"]["epoch_identifier"]="minute"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
 
+	jq '.app_state["oracle"]["params"]["tokens"][1]["asset_id"]="0xdac17f958d2ee523a2206206994597c13d831ec7_0x65"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
+	jq '.app_state["oracle"]["params"]["tokens"][1]["decimal"]="0"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
+
 	# custom epoch definitions can be added here, if required.
 	# see https://github.com/ExocoreNetwork/exocore/blob/82b2509ad33ab7679592dcb1aa56a7a811128410/local_node.sh#L123 as an example
```

Results
```yaml
exocored query bank balances exo17xpfvakm2amg962yls6f84z3kell8c5lgqz40d && sleep 60 && exocored query bank balances exo17xpfvakm2amg962yls6f84z3kell8c5lgqz40d

balances: []
pagination:
  next_key: null
  total: "0"

balances:
- amount: "20000000000000000000"
  denom: aexo
pagination:
  next_key: null
  total: "0"
```
